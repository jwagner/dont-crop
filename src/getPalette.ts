import { clusterHistogram } from './clusterHistogram';
import { Color } from './color';
import { greedyClusterMerge } from './greedyClusterMerge';
import { kmeans } from './kmeans';
import { labToSrgb, srgbToLabArray } from './lab';

export function getPalette(imageData: ImageData, numberOfColors: number): Color[] {
  const labData = srgbToLabArray(imageData.data);
  const k = Math.max(32, numberOfColors);
  const initialClusters = clusterHistogram(labData, 6, k);
  const clusters = kmeans(labData, initialClusters, 8).filter((c) => c.count > 0);
  return greedyClusterMerge(clusters, numberOfColors)
    .map(
      (cluster) => {
        const srgb = labToSrgb({ l: cluster.x, a: cluster.y, b: cluster.z });
        return [
          srgb.r * 255,
          srgb.g * 255,
          srgb.b * 255,
        ];
      },
    );
}
