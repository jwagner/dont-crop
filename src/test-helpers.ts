import { join } from 'path';
import sharp from 'sharp';
import { srgbToLabComposed } from './lab';

export async function getTestImageData(): Promise<ImageData> {
  const src = join(__dirname, '../tests/image.jpg');
  const { data, info } = await sharp(src).ensureAlpha()
    .toColorspace('srgb')
    .raw()
    .toBuffer({ resolveWithObject: true });
  const imageData: ImageData = {
    width: info.width,
    height: info.height,
    data: new Uint8ClampedArray(data.buffer),
  };
  return imageData;
}

export async function getLabTestData() {
  const imageData = await getTestImageData();
  return srgbToLabComposed(imageData.data);
}
