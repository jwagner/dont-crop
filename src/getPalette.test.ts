import { getPalette } from './getPalette';
import { getTestImageData } from './test-helpers';

describe('getPalette', () => {
  test('matches snapshot', async () => {
    const imageData = await getTestImageData();
    const actual = getPalette(imageData, 5);
    expect(actual).toMatchSnapshot();
  });
});
