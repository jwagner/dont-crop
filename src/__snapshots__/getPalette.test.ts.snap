// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPalette matches snapshot 1`] = `
Array [
  Array [
    245.47941244784906,
    35.38012362568841,
    38.769998350045114,
  ],
  Array [
    239.41387434330122,
    142.53473842464354,
    55.597169584742026,
  ],
  Array [
    48.5757696794788,
    176.37093542504869,
    155.72225005778623,
  ],
  Array [
    225.9497175004544,
    158.68869438908024,
    128.0015146932828,
  ],
  Array [
    114.26131572629866,
    177.77638243665783,
    218.73957466793604,
  ],
]
`;
